src/app/shared-services/inferences/engine/staypoint-engine/index.ts

```diff
@@ -0,0 +1,106 @@
+import { Inference } from 'src/app/model/inference'
+import { Trajectory } from 'src/app/model/trajectory'
+import {
+  IInferenceEngine,
+  InferenceDefinition,
+  InferenceResult,
+  InferenceResultStatus,
+  InferenceType,
+} from '../types'
+import { NightnessScoring } from '../../scoring/nightness-scoring'
+import { IInferenceScoring } from '../../scoring/types'
+import { WorkHoursScoring } from '../../scoring/work-hours-scoring'
+import {
+  inferHomeFromStayPointClusters,
+  inferWorkFromStayPointClusters,
+} from 'src/app/shared-services/staypoint/utils'
+import { StaypointService } from 'src/app/shared-services/staypoint/staypoint.service'
+
+export class StaypointEngine implements IInferenceEngine {
+  scorings: IInferenceScoring[] = [
+    new NightnessScoring(),
+    new WorkHoursScoring(),
+  ]
+
+  constructor(private staypointService: StaypointService) {}
+
+  private inputCoordinatesLimit = 100000
+
+  async infer(
+    trajectory: Trajectory,
+    inferences: InferenceDefinition[]
+  ): Promise<InferenceResult> {
+    if (trajectory.coordinates.length > this.inputCoordinatesLimit) {
+      return {
+        status: InferenceResultStatus.tooManyCoordinates,
+        inferences: [],
+      }
+    }
+
+    if (trajectory.coordinates.length === 0) {
+      return {
+        status: InferenceResultStatus.noInferencesFound,
+        inferences: [],
+      }
+    }
+
+    const daysInTrajectory = this.countDays(trajectory)
+    const weekDaysInTrajectory = this.countWeekDays(trajectory)
+
+    await this.staypointService.updateStayPoints(trajectory.type, trajectory.id)
+    const stayPoints = await this.staypointService.getStayPoints(trajectory.id)
+    const stayPointClusters =
+      await this.staypointService.computeStayPointClusters(stayPoints)
+
+    const inferenceResultsNested = inferences.map((i) => {
+      if (i.type === InferenceType.home) {
+        return inferHomeFromStayPointClusters(
+          stayPointClusters,
+          daysInTrajectory
+        )
+      }
+      if (i.type === InferenceType.work) {
+        return inferWorkFromStayPointClusters(
+          stayPointClusters,
+          weekDaysInTrajectory
+        )
+      }
+    })
+    // flatten and filter
+    const inferenceResults: Inference[] = [].concat
+      .apply([], inferenceResultsNested)
+      .filter((i) => i)
```

Comment:
What is the purpose of this `filter`? If I'm not mistaken, it should be without effect, right?

Reply:
this filters "falsey" values, like if there are no or undefined inferences.